internal/graph/reachableresources.go

```diff
@@ -0,0 +1,309 @@
+				Relation:  containingRelation.Relation,
+			}
+
+			// Otherwise, redispatch.
+			g.Go(crr.redispatch(
+				subCtx,
+				entrypoint,
+				stream,
+				&v1.DispatchReachableResourcesRequest{
+					ObjectRelation: req.ObjectRelation,
+					Subject:        rewrittenSubjectTpl,
+					Metadata: &v1.ResolverMeta{
+						AtRevision:     req.Revision.String(),
+						DepthRemaining: req.Metadata.DepthRemaining - 1,
+					},
+				},
+			))
+
+		case core.ReachabilityEntrypoint_TUPLESET_TO_USERSET_ENTRYPOINT:
+			containingRelation := entrypoint.ContainingRelationOrPermission()
+
+			// TODO(jschorr): Should we put this information into the entrypoint itself, to avoid
+			// a lookup of the namespace?
+			nsDef, ttuTypeSystem, err := crr.nsm.ReadNamespaceAndTypes(ctx, containingRelation.Namespace, req.Revision)
+			if err != nil {
+				return err
+			}
+
+			ttu := entrypoint.TupleToUserset(nsDef)
+			if ttu == nil {
+				return fmt.Errorf("found nil ttu for TTU entrypoint")
+			}
+
+			// Search for the resolved subject in the tupleset of the TTU. Note that we need to do so
+			// for both `...` as well as the subject's defined relation, as either is applicable in
+			// the tupleset (the relation is ignored when following the arrow).
+			relations := strset.New(tuple.Ellipsis, req.Subject.Relation)
+
+			for _, subjectRelation := range relations.List() {
+				isAllowed, err := ttuTypeSystem.IsAllowedDirectRelation(ttu.Tupleset.Relation, req.Subject.Namespace, subjectRelation)
+				if err != nil {
+					return err
+				}
+
+				if isAllowed != namespace.DirectRelationValid {
+					continue
+				}
+
+				ds := datastoremw.MustFromContext(ctx)
+				it, err := ds.ReverseQueryTuples(
+					ctx,
+					tuple.UsersetToSubjectFilter(&core.ObjectAndRelation{
+						Namespace: req.Subject.Namespace,
+						ObjectId:  req.Subject.ObjectId,
+						Relation:  subjectRelation,
+					}),
+					req.Revision,
+					options.WithResRelation(&options.ResourceRelation{
+						Namespace: containingRelation.Namespace,
+						Relation:  ttu.Tupleset.Relation,
+					}),
+				)
+				if err != nil {
+					return err
+				}
+				defer it.Close()
+
+				for tpl := it.Next(); tpl != nil; tpl = it.Next() {
+					if it.Err() != nil {
+						return it.Err()
+					}
+
+					rewrittenObjectTpl := &core.ObjectAndRelation{
+						Namespace: containingRelation.Namespace,
+						ObjectId:  tpl.ObjectAndRelation.ObjectId,
+						Relation:  containingRelation.Relation,
+					}
+
+					g.Go(crr.redispatch(
+						subCtx,
+						entrypoint,
+						stream,
+						&v1.DispatchReachableResourcesRequest{
+							ObjectRelation: req.ObjectRelation,
+							Subject:        rewrittenObjectTpl,
+							Metadata: &v1.ResolverMeta{
+								AtRevision:     req.Revision.String(),
+								DepthRemaining: req.Metadata.DepthRemaining - 1,
+							},
+						},
+					))
+				}
+			}
+
+		default:
+			panic(fmt.Sprintf("Unknown kind of entrypoint: %v", entrypoint.EntrypointKind()))
```

Comment:
so if we add a new kind of entrypoint, we can't do a rolling restart?

Reply:
Not until the new code is deployed; this is to ensure we don't accidentally miss handling a kind of entrypoint and therefore miss loading the data on older code; it is better IMO to panic and fail the API call than to return incomplete data